popularity-calculation-script/src/popularity_scraper.py

```diff
@@ -0,0 +1,111 @@
+import json
+import re
+from urllib.request import urlopen
+
+import pandas as pd
+
+
+class PopularityScraper:
+    def __init__(self, url: str, filename: str) -> None:
+        self.url = url
+        self.filename = filename
+        self.df = pd.DataFrame()
+
+    def download_data_from_kiwix_url(self, label="download.kiwix.org") -> dict:
+        """Get the relevent dictionery from the list of dictioneries from the data
+
+        Args:
+            label (str, optional): Label key of dictionary. Defaults to "download.kiwix.org".
+
+        Returns:
+            dict: the inner dictionary with the relevent label
+        """
+        response = urlopen(self.url)
+        data_json = json.loads(response.read())
+        for i in data_json:
+            i["label"] == label
+            subtable = i["subtable"]
+            self.df = pd.DataFrame(subtable)
+            break
+
+    def remove_nonzim_data_from_df(self):
+        # clean the data; remove rows that don't have "zim" substring in the label column
+        ## refer: https://stackoverflow.com/questions/17097643/search-for-does-not-contain-on-a-dataframe-in-pandas
+        self.df = self.df.drop(
+            self.df[~self.df.label.str.endswith(".zim.torrent")].index
+        )
+
+    def add_zim_name_column(self):
+        for index, row in self.df.iterrows():
+            self.df.at[index, "zim"] = self.extract_zim_name_from_label(row["label"])
+
+    def extract_zim_name_from_label(self, label: str) -> str:
+        """Get the zim name from the label without the slash prefixes and date
+
+        Args:
+            label (str): Label of the row
+
+        Returns:
+            str: zim name
+        """
+        pattern = r"^.*\/(.*)\.zim.*$"
+        match = re.search(pattern, label)
+
+        # removing year and month
+        match_split = match.group(1).split("_")
```

Comment:
Even though you filtered entries not ending in `.zim.torrent` in other method, it would be safer to ensure match succeeded before accessing the group